File: src/Lavoisier/Protocol/Gateway.hs
{- | The 'Gateway' contract: a frontend that serves the shared agent. Ported from Rust
@lvz-protocol@ @gateway.rs@ (@Arc\<dyn Gateway\>@ → a record of functions).
-}
module Lavoisier.Protocol.Gateway (
    GatewayError (..),
    Gateway (..),
)
where

import Data.Text (Text)

import Lavoisier.Protocol.Agent (AgentHandle)

-- | Errors a gateway can fail with.
data GatewayError
    = -- | Could not bind\/authenticate (a genuine config error; surfaces immediately).
      GEBind Text
    | -- | A transient I\/O failure (retryable).
      GEIo Text
    | -- | A protocol-level failure.
      GEProtocol Text
    deriving stock (Eq, Show)

{- | A gateway, as a record of functions. 'gatewayServe' runs until shutdown, driving the shared
'AgentHandle'.
-}
data Gateway = Gateway
    { gatewayName ∷ Text
    , gatewayServe ∷ AgentHandle → IO (Either GatewayError ())
    }

File: crates/lvz-protocol/src/gateway.rs
//! The [`Gateway`] contract [Hermes tier]. A gateway is a frontend/channel (CLI, HTTP,
//! Matrix) that drives the shared agent via an [`AgentHandle`](crate::AgentHandle).
//! Adding a channel means one new `lvz-gw-*` crate; the agent core stays unaware of it
//! (§5.5, §7).

use std::sync::Arc;

use async_trait::async_trait;

use crate::agent::AgentHandle;

/// A frontend that runs an event loop, translating inbound requests into agent turns and
/// rendering outbound [`Event`](crate::Event)s back to its channel.
#[async_trait]
pub trait Gateway: Send + Sync {
    /// Identifier for logs/telemetry (e.g. `cli`, `http`, `matrix`).
    fn name(&self) -> &str;

    /// Run the gateway's serve loop until shutdown, dispatching to the shared `agent`.
    async fn serve(self: Arc<Self>, agent: Arc<dyn AgentHandle>) -> Result<(), GatewayError>;
}

/// Errors surfaced by a gateway's serve loop.
#[derive(Debug, thiserror::Error)]
pub enum GatewayError {
    /// Failed to bind/listen/connect the channel.
    #[error("bind error: {0}")]
    Bind(String),

    /// Channel-level I/O failure during serving.
    #[error("io error: {0}")]
    Io(String),

    /// Inbound payload could not be parsed into a turn.
    #[error("protocol error: {0}")]
    Protocol(String),
}
